librad/src/git/replication.rs

```diff
@@ -106,194 +128,599 @@ pub fn replicate<Addrs>(
-                        source: e,
-                    })?
-                    .symbolic_ref::<_, PeerId>(
-                        Reference::rad_delegate(Namespace::from(&urn), &delegate_urn),
-                        Force::False,
-                    )
-                    .create(storage.as_raw())
-                    .and(Ok(()))
-                    .or_matches(is_exists_err, || Ok(()))
-                    .map_err(|e: git2::Error| Error::Store(e.into()))
+
+                Ok(delegations)
             },
         }?;
-    }
 
-    let proj = identities::project::verify(storage, &urn)?
-        .ok_or(Error::MissingIdentity)?
-        .into_inner();
-
-    // Create `rad/id` here, if not exists
-    ensure_rad_id(storage, &urn, proj.content_id)?;
-
-    // Make sure we track any direct delegations
-    for key in proj
-        .delegations()
-        .iter()
-        .direct()
-        .filter(|&key| key != local_peer_id.as_public_key())
-    {
-        tracking::track(storage, &urn, PeerId::from(*key))?;
+        // Create `rad/id` here, if not exists
+        adopt_latest(storage, &urn, delegations)?;
+        Ok(())
     }
 
-    Ok(proj
-        .doc
-        .delegations
-        .into_iter()
-        .indirect()
-        .map(|id| id.urn()))
+    /// Adopt the `rad/id` that has the most up-to-date commit from the set of
+    /// `Person` delegates.
+    #[allow(clippy::unit_arg)]
+    #[tracing::instrument(level = "trace", skip(storage), err)]
+    pub fn adopt_latest(
+        storage: &Storage,
+        urn: &Urn,
+        delegates: BTreeSet<PeerId>,
+    ) -> Result<(), Error> {
+        let persons = NonEmpty::from_vec(
+            delegates
+                .into_iter()
+                .flat_map(|peer| {
+                    let urn =
+                        unsafe_into_urn(Reference::rad_id(Namespace::from(urn)).with_remote(peer));
+                    identities::person::get(storage, &urn).ok().flatten()
+                })
+                .collect(),
+        );
+        let tip = match persons {
+            Some(persons) => identities::person::latest_tip(storage, persons).map_err(Error::from),
+            None => Err(Error::MissingIdentities(urn.clone())),
+        }?;
+        ensure_rad_id(storage, urn, tip.into())
+    }
 }
 
-#[allow(clippy::unit_arg)]
-#[tracing::instrument(level = "trace", skip(storage), err)]
-fn ensure_rad_id(storage: &Storage, urn: &Urn, tip: ext::Oid) -> Result<(), Error> {
-    identities::common::IdRef::from(urn)
-        .create(storage, tip)
-        .map_err(|e| Error::Store(e.into()))
+mod project {
+    use super::*;
+
+    #[derive(Clone, Debug)]
+    pub struct DelegateView {
+        pub urn: Urn,
+        pub delegate: VerifiedPerson,
+        pub project: VerifiedProject,
+    }
+
+    /// Process the setup of a `Project` by:
+    ///   * Verifying the identity
+    ///   * Ensuring there are no forks between the remote we are replicating
```

Comment:
This is a good point, and it's the switch in perspective I needed for improving a bit of logic that seems a bit "corner-case"-y to me. That is this scenario where we're adding the remote to the set regardless https://github.com/radicle-dev/radicle-link/pull/478/files#diff-fe1d1a45efacfdf43a7b387a93ab0db0b6eeab3e68b613e5f190fcbcdafb6ff0R271-R273.

The last thing to solve to remove this is the case of getting the state of the delegates https://github.com/radicle-dev/radicle-link/pull/478/files#diff-fe1d1a45efacfdf43a7b387a93ab0db0b6eeab3e68b613e5f190fcbcdafb6ff0R575. In the case of cloning we're relying on the remote's view to get setup the delegates in our monorepo. As you pointed out above, we shouldn't find the remote interesting (unless we track them). So in the question is **who** do we look at to get the (possibly) new state of the delegates? 

I'm thinking it's pretty arbitrary, we could just pick _one_ delegate remote to understand the new state of the world. How does that sit with you?

Reply:
> So in the question is **who** do we look at to get the (possibly) new state of the delegates?

You mean for the fetch case? This is where the TODO at the end of `replicate` kicks in: we know _some_ set of delegates, which we'd all update. Because they bring their own delegates (`rad/ids/*`), it will be possible to verify updates without knowing any **new** delegates yet. We can also check if the known set agrees (doesn't fork), and determine a most-recent one (to which we can update our own `rad/id`, **unless** we are ourselves part of the delegates set, in which case we need to prompt the user ... somehow).

However, if new delegates where introduced, we now need to track them. And technically also fetch, although this could be deferred to later.

Reply:
> You mean for the fetch case?

Yes, exactly. 

> This is where the TODO at the end of replicate kicks in: we know some set of delegates, which we'd all update.

Ahhh, I see I thought that was something to do with I/O interactions and being out of sync.

This PR is making an attempt to solve this case. I'll try to lay out the steps it takes:
1. Peek at known delegates and tracked peers
2. Choose one of those peers, the `trustee`, to get the delegate views -- doing the adopt and track dance
3. Ensure there's no forking against our view of the project
4. Track direct peers
5. Fetch signed refs
6. Track the signed ref peers
7. Adopt the latest 
8. Prune any delegates or tracked peers that were removed

The `trustee` part was what I was getting at in my question, and I added it in the latest commit https://github.com/radicle-dev/radicle-link/pull/478/commits/c8d33950536a6ec59bcb0b21ede86ad11930ec6d.

What you're saying above seems to be that we don't need step 2. in that list. We just go off of the delegates that we know about. Seems like I just need to rework that logic to work with the known delegates and then extend from there with the new/old delegate case (with the tricky case of being a delegate yourself).

Reply:
This should be closer to what we want https://github.com/radicle-dev/radicle-link/pull/478/commits/5414f964ac8ddd4e46668e250eac4318f7b71370. We only fetch what we know, and I added an `enum` for signaling when we're in the case of being a delegate and there's a mismatch between the latest tip and our tip. However, I defaulted to saying we're "behind", but I think we could be "ahead" as well. I still don't have a good mental model for this **tree** and **commit** business. I'd appreciate a session on this to flesh it out in my head a bit more if you have the time. I'm looking over your comments and the docs to see if I can piece it together.